rig/machine_control/machine_controller.py

```diff
@@ -1739,34 +1807,169 @@ def get_machine(self, x=0, y=0, default_num_cores=18):
+
+        Returns
+        -------
+        [\
+        :py:class:`rig.place_and_route.constraints.ReserveResourceConstraint`,\
+        ...]
+            A set of place-and-route constraints which reserves non-idle cores
+            and SDRAM. The resource types given in the ``core_resource`` and
+            ``sdram_resource`` arguments will be reserved accordingly.
+        """
+        if _all_chips_info is not None:
+            chips_info = _all_chips_info[2]
+        else:
+            chips_info = {}
+
+        constraints = []
+
+        for x, y in machine:
+            chip_info = chips_info.get((x, y))
+            if chip_info is None:
+                chip_info = self.get_chip_info(x, y)
+
+            # Construct a minimal set of ReserveResourceConstraints which
+            # reserve the cores which are not in the idle state.
+            cur_reservation = None
+            for state, core in zip(chip_info.core_states,
+                                   range(machine[(x, y)][core_resource])):
+                if state == consts.AppState.idle:
+                    if cur_reservation is not None:
+                        constraints.append(cur_reservation)
+                    cur_reservation = None
+                else:
+                    if cur_reservation is None:
+                        cur_reservation = ReserveResourceConstraint(
+                            core_resource,
+                            slice(core, core + 1),
+                            (x, y))
+                    else:
+                        cur_reservation.reservation = slice(
+                            cur_reservation.reservation.start,
+                            core + 1)
+            if cur_reservation is not None:
+                constraints.append(cur_reservation)
+
+            # Reserve enough SDRAM such that the amount remaining will be the
+            # size of the largest free block in the heap.
+            total_sdram = machine[(x, y)][sdram_resource]
+            avail_sdram = chip_info.largest_free_sdram_block
+            constraints.append(ReserveResourceConstraint(
+                sdram_resource,
+                slice(avail_sdram, total_sdram),
+                (x, y)))
+
+            # Reserve enough SRAM such that the amount remaining will be the
+            # size of the largest free block in the heap.
+            total_sram = machine[(x, y)][sram_resource]
+            avail_sram = chip_info.largest_free_sram_block
+            constraints.append(ReserveResourceConstraint(
+                sram_resource,
+                slice(avail_sram, total_sram),
+                (x, y)))
+
+        return constraints
+
+    def get_machine_and_resource_constraints(self, x=0, y=0,
+                                             default_num_cores=18):
+        """An efficient wrapper around both :py:meth:`.get_machine` and
+        :py:meth:`.get_resource_constraints`.
```

Comment:
"Get a representation of the machine and constraints upon the resources which may be used."... or similar